File: pengsoft-support/pengsoft-support-biz/src/main/java/com/pengsoft/support/biz/service/TreeEntityService.java
package com.pengsoft.support.biz.service;

import com.pengsoft.support.domain.entity.TreeEntityImpl;
import com.pengsoft.support.domain.entity.TreeEntity;
import com.querydsl.core.types.Predicate;
import org.springframework.data.domain.Sort;

import java.io.Serializable;
import java.util.List;

/**
 * The service interface of {@link TreeEntityImpl}.
 *
 * @author dang.peng@pengsoft.com
 * @since 1.0.0
 */
public interface TreeEntityService<T extends TreeEntity<T, ID>, ID extends Serializable> extends EntityService<T, ID> {

    /**
     * Returns all entities under the parent specified in {@link Predicate}
     *
     * @param predicate {@link Predicate}, contains parent parameter, if not parent is null.
     * @param sort      {@link Sort}
     */
    List<T> findAllByParent(Predicate predicate, Sort sort);

    /**
     * Returns all entities under the parent specified in {@link Predicate}, excluding the given node and its children
     * nodes.
     *
     * @param self      The excluded node.
     * @param predicate {@link Predicate}, contains parent parameter, if not parent is null.
     * @param sort      {@link Sort}
     */
    List<T> findAllExcludeSelfAndItsChildrenByParent(T self, Predicate predicate, Sort sort);

    /**
     * Returns all entities, excluding the given node and its children nodes.
     *
     * @param self      The excluded node.
     * @param predicate {@link Predicate}, contains parent parameter, if not parent is null.
     * @param sort      {@link Sort}
     */
    List<T> findAllExcludeSelfAndItsChildren(T self, Predicate predicate, Sort sort);

    /**
     * Returns all entities that their 'parentIds' starts with the given value.
     *
     * @param parentIds The entity's parentIds
     */
    List<T> findAllByParentIdsStartsWith(String parentIds);

}

File: pengsoft-app/pengsoft-app-admin/src/app/services/commons/tree-entity.service.ts
import { EntityService } from './entity.service';
import { HttpService } from './http.service';
import { HttpOptions } from './http-options';

export abstract class TreeEntityService extends EntityService {

    constructor(protected http: HttpService) { super(http); }


    findAllByParent(parent: any, params: any, options: HttpOptions): void {
        const url = this.getApiPath('find-all-by-parent');
        options.params = Object.assign({}, params);
        if (parent && parent.id) {
            options.params['parent.id'] = parent.id;
        }
        this.http.request('GET', url, options);
    }

    findAllExcludeSelfAndItsChildrenByParent(parent: any, self: any, params: any, options: HttpOptions): void {
        const url = this.getApiPath('find-all-exclude-self-and-its-children-by-parent');
        options.params = Object.assign({}, params);
        if (parent && parent.id) {
            options.params['parent.id'] = parent.id;
        }
        if (self && self.id) {
            options.params['self.id'] = self.id;
        }
        this.http.request('GET', url, options);
    }

    findAllExcludeSelfAndItsChildren(self: any, params: any, options: HttpOptions): void {
        const url = this.getApiPath('find-all-exclude-self-and-its-children');
        options.params = Object.assign({}, params);
        if (self && self.id) {
            options.params['self.id'] = self.id;
        }
        this.http.request('GET', url, options);
    }

}
